telemetry/otel.go

```diff
@@ -0,0 +1,234 @@
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+var defaultStartOnce sync.Once
+
+const (
+	defaultOtelAgentAddr = "0.0.0.0:4317"
+)
+
+type (
+	meterProviderFunc  func(context.Context, *resource.Resource) (*metric.MeterProvider, error)
+	tracerProviderFunc func(context.Context, *resource.Resource) (*trace.TracerProvider, error)
+)
+
+// Start initializes automatic instrumentation for exporting OTLP metrics and traces
+// to an OpenTelemetry collector agent.
+func Start(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newOTLPTraceProvider, newOTLPMeterProvider)
+	})
+	return
+}
+
+// StartWithStdoutExport initializes automatic instrumentation for exporting OTLP
+// metrics and traces to stdout (e.g. Cloud Logging if using Cloud Run).
+//
+// This is intended for testing and debugging metrics and traces before exporting
+// them to an OpenTelemetry collector agent.
+func StartWithStdoutExport(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newStdoutTraceProvider, newStdoutMeterProvider)
+	})
+	return
+}
+
+func start(ctx context.Context, newTracer tracerProviderFunc, newMeter meterProviderFunc) (shutdown func(context.Context) error, err error) {
+	var shutdownFuncs []func(context.Context) error
+
+	shutdown = func(ctx context.Context) error {
+		var err error
+		for _, fn := range shutdownFuncs {
+			err = errors.Join(err, fn(ctx))
+		}
+		shutdownFuncs = nil
+		return err
+	}
+
+	if err := runtime.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start runtime instrumentation: %w", err)
+	}
+	// Set up some automatic attributes regarding deployment environment, service,
+	// GCP, etc to attach to telemetry.
+	res, err := resource.New(
+		ctx,
+		resource.WithFromEnv(),
+		resource.WithProcess(),
+		resource.WithDetectors(gcp.NewDetector()),
+		resource.WithTelemetrySDK(),
+		resource.WithHost(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create opentelemetry resource: %w", err)
+	}
+
+	// Set up propagator to W3C trace context, a common specification for
+	// distributed tracing. See https://opentelemetry.io/docs/concepts/context-propagation.
+	prop := newPropagator()
+	otel.SetTextMapPropagator(prop)
+
+	// Set up tracing.
+	tracerProvider, err := newTracer(ctx, res)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up tracing provider: %w", err)
+	}
+	shutdownFuncs = append(shutdownFuncs, tracerProvider.Shutdown)
+	otel.SetTracerProvider(tracerProvider)
+
+	// Set up monitoring.
+	meterProvider, err := newMeter(ctx, res)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up meter provider: %w", err)
```

Comment:
Not sure what the most idomatic solution is here, but unless the client calls shutdown function even in the case of err, we will leak the tracer provider.

Reply:
Why would there be a leak of the tracer provider?

It's handled here, no?
https://github.com/abcxyz/pkg/pull/357/files/b322e73f4443fe3be006b0f0c364c160931f1071#diff-bc1c8ec1ff1a3119d698617846d3dd1c221200f61df98429eb0ad39fabe7d285R130

Reply:
I guess what I am saying is, it needs to be documented what the behavior is upon an error being returned.

Is the caller expected to call `shutdown()`?

@drevell Probably would be the better person to say what is more idiomatic. My expectation would be if you return error you shouldn't expect users to call `shutdown()`, but I'm not as familiar with go style.

Reply:
Obsolete. Now the tracer and meter are started separately. PTAL